src/DurableTask.Netherite.AzureFunctions/NetheriteProviderFactory.cs

```diff
@@ -114,15 +118,31 @@ NetheriteOrchestrationServiceSettings GetNetheriteOrchestrationServiceSettings(s
                 netheriteSettings.HubName = taskHubNameOverride;
             }
 
+            // connections for Netherite are resolved either via an injected custom resolver, or otherwise by resolving connection names to connection strings
+            var connectionResolver = this.serviceProvider?.GetService<DurableTask.Netherite.ConnectionResolver>()
```

Comment:
I think there is an opportunity to follow DI/IoC with the connection resolver, but still make it ultimately "optional" for customers. The idea would be to move this kind of logic to the DI setup code. I'll have to take a closer look at what existing DI setup code Netherite has, but I think we can definitely get something working.

For example, is `INameResolver` even used directly here, or is it only used within the `ConnectionResolver`? If it is only used within the connection resolver, we can do the following:

1. Add a 2nd constructor which takes in `ConnectionResolver` and _no_ `INameResolver`.
2. The ctor which still has `INameResolver` just calls that new ctor with the `new ConnectionNameToConnectionStringResolver((name) => nameResolver.Resolve(name))` as the connection resolver.
    - No longer have a `INameResolver` field. Which is why it is important INameResolver is not used directly anymore. If we still need it directly, will need to rethink this aproach.
4. Add `[ActivatorUtilitiesConstructor]` to the new connection resolver ctor (so it is used for DI).
5. Update the DI setup code to use `TryAddSingleton<ConnectionResolver>` with a factory which will pull `INameResolver` from the service container and return `new ConnectionNameToConnectionStringResolver((name) => nameResolver.Resolve(name))` as well.
    - at this point, maybe a constructor for `ConnectionNameToConnectionStringResolver` which takes in `INameResolver` may be nice.

What we now have is no observable difference between old and new. However, customers can now inject `ConnectionResolver` to the DI container and it will be consumed as expected, but we have maintained DI/IoC principals.

Reply:
> However, customers can now inject ConnectionResolver to the DI container 

I don't quite understand how they can inject a ConnectionResolver if we already injected our own during our own setup

Reply:
Empirically, it does seem to work... 
- The TokenCredentialDF.FunctionStartup calls `builder.Services.AddSingleton<DurableTask.Netherite.ConnectionResolver, MyConnectionResolver>();`
- The DurableTask.Netherite.AzureFunctions.NetheriteProviderStartup calls `builder.Services.TryAddSingleton<ConnectionResolver, NameResolverBasedConnectionNameResolver>();`

and magically the right things happen. Did I just get lucky or is the DI smart enough to know which one should take precedence?

Reply:
> If we still need it directly, will need to rethink this aproach

I still need it, but that does not really change anything I think. I just leave it there, and go ahead with the rest of the suggestions.

Reply:
What is happening is `builder.Services.TryAddSingleton<ConnectionResolver, NameResolverBasedConnectionNameResolver>();` will first check a service for `ConnectionResolver` is already registered, only adding `NameResolverBasedConnectionNameResolver` if no `ConnectionResolver` is already registered. This gives end users control over the which instance is registered: if they run before you, they can just call `AddSingleton`. If they run after, they will first need to remove your service registration then add their own (this is possible, `IServiceCollection` is a fully mutable collection with add/remove semantics).

Reply:
Apparently it is not necessary for users to make a distinction of what order the code is running in. 

I stepped through the sample in the debugger to see why this is working...  what I observed is that the user code is running after me and calls AddSingleton, which I would have expected to cause some problem. Instead, everything ends up working as desired. The debugger shows that as a result, immediately after this step, there are two services of this type in the collection:
![image](https://user-images.githubusercontent.com/6415760/200046284-4942b0f4-388c-4ad9-8a3e-75bc1261ec78.png)
And when the NetheriteProviderFactory is constructed by DI, it is given the second one of these services, which is the desired result. 

It is a bit strange, but perhaps no need to worry since it is working correctly.